### ISSUE TITLE
Gmail web UI does not "fold" HTML citations generated by org-mime

### ISSUE THREAD

**Issue by elizagamedev** [org-mime/org-mime#78]
GMail shows a small ellipsis for citations in emails composed by the web UI. This, unfortunately, does not work with org-mime HTML emails. Below is an example of the difference in generated HTML output.

GMail-generated message, reformatted for readability
```html
<div class="gmail_quote">
  <div dir="ltr" class="gmail_attr">On Sat, May 14, 2022, 6:10 PM xxx &lt;<a href="mailto:xxx@xxx">xxx@xxx</a>&gt; wrote:
    <br>
  </div>
  <blockquote class="gmail_quote" style="margin:0 0 0 .8ex;border-left:1px #ccc solid;padding-left:1ex">
    <!-- ... -->
  </blockquote>
</div>
```

org-mime generated message, reformatted for readability
```html
<p>
  On Sat, May 14 2022, xxx wrote:
</p>

<p>
  <blockquote class="gmail_quote" style="margin:0 0 0 .8ex;border-left:1px #ccc solid;padding-left:1ex">
    <!-- ... -->
  </blockquote>
</p>
```

I assume the `<div class="gmail_quote">` might fix things, but I'm not certain.

GMail detects the citations of plaintext messages just fine, whether top-posting or bottom-posting, with signature and without, and it's quite forgiving of the formatting of the attribution line. The Fastmail web UI detects org-mime attributions correctly, so it's extra frustrating that GMail is so picky with HTML mail. That said, I understand if getting the GMail web UI is out of scope for this project.

**Comment by redguardtoo** [comment]
Could you give me an minimum example to reproduce the issue? I forget how to quote the original mail in gnus.

**Comment by elizagamedev** [comment]
```
This ought to do it! Generated from replying to the GitHub notification email and copying the text into a GitHub issue comment. I'll also mail this to my gmail account with org-mime, take a screenshot, and post another comment.

On Sat, May 14 2022 at 23:30 -07, Chen Bin <notifications@github.com> wrote:

> Could you give me an minimum example to reproduce the issue? I forget how to quote the original mail in gnus.

-- 
Eliza
```

**Comment by elizagamedev** [comment]
Awesome! I just tested the commit out for myself, and I have a few questions/concerns. Sorry if this is just a work in progress feature that I'm overly eager about, I'm just very excited to use Emacs at work to compose mail :)

1. Is the intent that a string such as `>>>>> == citation` get automatically prefixed before the citation by prepending it to `message-citation-line-format`, or is this already an Emacs message construct that just isn't part of my existing workflow? (I use notmuch)
2. As a followup to that, recursive quotes no longer seem to be supported (for example, a nested citation with `>>>` is simply translated to a series of `&gt;&gt;`)
3. It seems like in order to match the replies formatted by GMail, the inline CSS is necessary, otherwise you get the extreme indentation like in the screenshot you posted.

**Comment by madalu** [comment]
Notmuch uses message-mode for composing emails (same as Gnus). However, the default behavior in message mode is to insert a simple citation line: "John Doe (john@example.com) writes:" This is determined by the variable message-citation-line-function, which is set to message-insert-citation-line by default. One only gets the fancier formatted line "On... wrote:" if one sets message-citation-line-function to message-insert-formatted-citation-line.

**Comment by madalu** [comment]
As @elizagamedev noted with &qt; entities and nested quoting, these changes introduced several regressions. I've opened a separate issue here: #79.

**Comment by elizagamedev** [comment]
Yes, as @madalu says. When I respond to an email, this is what it looks like.

![reply](https://user-images.githubusercontent.com/4576666/169634295-66ec0e04-2076-4c84-be7f-b5d84729c0bf.png)

I'm curious about what your gnus setup looks like in comparison, since I've never seen a citation that looked like `>>>>> ==` or similar. I appreciate how quickly you responded to my issue but I do agree that the regressions introduced by this change are too much and should be reverted for now.

**Comment by redguardtoo** [comment]
Sorry. I need rollback the code to previous stable version. I will set up a branch to work out this new feature. It's more complicated than I expected. I need create some unit tests first.

**Comment by redguardtoo** [comment]
Leave this ticket open. I've set up a new branch `gmail-style`,  I may revisit this ticket in the future,

some notes,
- extract mail body and quoted mail before it's htmlized
- commit 10098b001c3926f0fb128c85fc64d9016a195868 implemented nest quoted algorithm, need be studied first
- in gmail dom tree for quoted mail is like,
```html
,<div class="im">
  <div class="gmail_quote">
      <div class="gmail_attr">
      </div>
      <blockquote class="gmail_quote">
      </blockquote>
  </div>
</div>
```
- see the `show original mail` menu in gmail web ui.

**Comment by krisbalintona** [comment]
> Leave this ticket open. I've set up a new branch `gmail-style`, I may revisit this ticket in the future,
> 
> some notes,
> 
>     * extract mail body and quoted mail before it's htmlized
> 
>     * commit [10098b0](https://github.com/org-mime/org-mime/commit/10098b001c3926f0fb128c85fc64d9016a195868) implemented nest quoted algorithm, need be studied first
> 
>     * in gmail dom tree for quoted mail is like,
> 
> 
> ```
> ,<div class="im">
>   <div class="gmail_quote">
>       <div class="gmail_attr">
>       </div>
>       <blockquote class="gmail_quote">
>       </blockquote>
>   </div>
> </div>
> ```
> 
>     * see the `show original mail` menu in gmail web ui.

Hi! I just found this issue since I've been trying to figure out this problem myself. I actually use `org-msg`, but the problem with Gmail formatting applies there too.

I've been able to get `org-msg` to wrap the citation line in a div with class "gmail_attr", the entire citation in a div with class "gmail_quote", and the citation text in a blockquote with class "gmail_quote" --- just as you described above. However, this doesn't always work in folding/buttonizing the citation. I haven't been able to figure out exactly when it doesn't work, but it definitely works in reply to very simple, plain-text only emails. When it fails, it won't buttonize/fold the citation but it will recognize it as such and set the foreground color of the text to purple (as it does when you expand the folded citation).

I was wondering whether anyone has any hints on what else Gmail does to ensure citations are folded.  Inspecting the raw mail files, as far as I can tell, the only difference the value of the style properties.

**Comment by redguardtoo** [comment]
> Leave this ticket open. I've set up a new branch `gmail-style`, I may revisit this ticket in the future,
You could send me a PR if you have tested it in gmail and there is no regression.
